text/2020-12-16-pd-rbac.md

```diff
@@ -0,0 +1,150 @@
+# RBAC Support for PD
+
+## Summary
+
+Here we propose adding RBAC to PD so that all requests to PD must be authenticated,
+and clients may only do actions they are explicitly authorized to perform.
+
+This RFC is a partial implementation of [tikv/8621](https://github.com/tikv/tikv/issues/8621)
+that describes a way to support RBAC-based data access control for TiKV and PD.
+
+## Motivation
+
+PD server exposes its APIs to the network via HTTP or gRPC endpoints.
+
+Currently, clients knowing the PD address have full access to PD APIs.
+This poses a threat to the PD and TiKV cluster.
+
+If a PD server is exposed or a client is compromised, an attacker may freely
+read the status of the PD cluster and perform destructive actions.
+
+And in some cases, we may want to authorize a client (most likely some utility)
+access to a subset of PD APIs, which can’t be realized without an auth system.
+
+Once RBAC support is added into PD, we may assign specific permissions to a client,
+and make sure no client may perform actions beyond its need, minimizing the potential
+outcome of a security breach.
+
+Utilities requiring direct access to PD may gain limited access to it with a user
+created earlier by an administrator.
+
+## Detailed design
+
+### Authentication
+
+- PD stores `User`s in a KV database (etcd).
+  - Username: string
+  - Hash: string (sha256)
+  - Roles: a set of strings (role names)
+- A client (TiDB/TiKV/pd-ctl/…) must provide a pair of username and password to
+  gain access to PD APIs (including HTTP and gRPC endpoints).
+  - Clients using HTTP APIs should send requests with a basic authentication header.
```

Comment:
Can users always use https to contact PD? Using http for this would be insecure.

Reply:
That's a good idea! We may enforce TLS when authentication is enabled.